### PR TITLE
Rewrite git history

### PR DESCRIPTION
**Pull request by andiradulescu**
This tackles https://github.com/commaai/openpilot/issues/28840 and is a continuation of https://github.com/commaai/openpilot/pull/30824.

This PR addresses `master`:

- [x] prevent new large files https://github.com/commaai/openpilot/pull/30839
- [x] move old objects, binary objects and big text files to LFS https://github.com/commaai/openpilot/pull/31562/files#diff-35f170270f2172a621c54f4e4ceb7f9675b74a90eb359b376202f6a80c819aa2R231
- [x] prepend the devel release history - from `v0.7.1`
- [x] tags - all tagged off master
- [x] one liner for external contributors to create a new branch to open a new PR on the new history (`git cherry-pick A^..B` where A is the first commit and B is the last commit, from their old PR)
- [x] validation of the new history
- [x] archive of old history - if over 2GB, `git push` fails
- [x] commits will contain their old commit id, Former-commit-id: b38c580c2e687401fc819520ea038653873e9996
- [x] rebase all branches on upstream - except ones that can't be done automatically and prebuilt branches
- [x] deal with PRs - PRs stay open if the rest of the branches are also rewritten
- [ ] test the script at least a couple of times

Continued in https://github.com/commaai/openpilot/pull/32955

### CODE REVIEW

**Comment by jnewb1** [comment]
```20.83 MiB``` ❤️

image

**Comment by andiradulescu** [comment]
Good eye @jnewb1 - yes, it's very small ❤️

GitLab LFS is 6.3 GiB though, but that's not an issue. Since when you cloned openpilot-tiny, only 250MB of LFS files are downloaded.

I uploaded the LFS on GitLab (as you do).

File: scripts/bfg.sh [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+# this is needed after lfs import
+git reflog expire --expire=now --all
+git gc --prune=now --aggressive
+
+# don't delete external and phonelibs anymore (?!)
+# git-filter-repo doesn't seem to need reflog and gc after the command
+# /tmp/git-filter-repo --invert-paths --path external/ --path phonelibs/
+
+# check the git-filter-repo analysis again - can be found in the repo root/filter-repo/analysis
+/tmp/git-filter-repo --force --analyze
```
Comment (adeebshihadeh) [inline scripts/bfg.sh:87]: i'm most interested in knowing the top remaining files and their sizes. does this have it?

File: scripts/bfg.sh [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+cp -r $SRC $OUT
+
+cd $OUT
+
+# remove all tags
+# TODO: we need to keep the tags, WIP on how we actually do it
+# git tag -l | xargs git tag -d
+
+# remove all non-master branches
+# TODO: need to see if we "redo" the other branches (except master, master-ci, devel, devel-staging, release3, release3-staging, dashcam3, dashcam3-staging, testing-closet*, hotfix-*)
```
Comment (adeebshihadeh) [inline scripts/bfg.sh:70]: would be nice to automatically rebase the other branches on the repo, but guess we can split that out into a second script that runs on the generated repo from here

File: scripts/delete.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import git_filter_repo as gfr
+
```
Comment (adeebshihadeh) [inline scripts/delete.py:2]: feel free to delete this if it's unused

**Comment by andiradulescu** [comment]
My strategy (like it or hate it):

1. push new master as “main” branch (new branch)
2. put a github action on master on every push (to master) to take the last commit that was pushed to master and cherry pick it on “main”
3. make “main” the main branch of the github repo
4. everyone works as they worked before, just that when they start a new feature they should now start from “main”
5. for some time there will be PRs opened for “main” and PRs open for “master”
6. we set a deadline for 30 days for everyone to move to “main”
7. when we reach the deadline we close all opened PRs to “master” and remove the “master” branch (and push it to archive repo)
8. clean all other branches we don’t need anymore
9. clean the repo of anything else that might need to be cleaned so that a new clone is 20MB not 3GB

The strategy takes 30 days to apply, but shouldn’t disrupt anyone in the 30 days.

After the 30 days, will be disruptive for users that need to move forcefully on main.

Didn’t do this strategy before. It’s a new idea.

**Comment by devtekve** [comment]
> My strategy (like it or hate it):
> 
> 1. push new master as “main” branch (new branch)
> 2. put a github action on master on every push (to master) to take the last commit that was pushed to master and cherry pick it on “main”
> 3. make “main” the main branch of the github repo
> 4. everyone works as they worked before, just that when they start a new feature they should now start from “main”
> 5. for some time there will be PRs opened for “main” and PRs open for “master”
> 6. we set a deadline for 30 days for everyone to move to “main”
> 7. when we reach the deadline we close all opened PRs to “master” and remove the “master” branch (and push it to archive repo)
> 8. clean all other branches we don’t need anymore
> 9. clean the repo of anything else that might need to be cleaned so that a new clone is 20MB not 3GB
> 
> The strategy takes 30 days to apply, but shouldn’t disrupt anyone in the 30 days.
> 
> After the 30 days, will be disruptive for users that need to move forcefully on main.
> 
> Didn’t do this strategy before. It’s a new idea.

If I can add some input, for the step 6 I’d only apply it to new PRs, otherwise it gives a 30day timeout for existing WIP which might be a bit tight. New PRs after the deadline should be rejected in favor of PRs to main

**Comment by adeebshihadeh** [comment]
RE: rollout, I'm not interested in dealing with a slow rollout, so we'll do it quickly and put some effort into rebasing PRs automatically for people.

**Comment by andiradulescu** [comment]
Small update here: rebasing `master` onto `v0.7.1` took more than expected until I found the proper hacks. Check `handle_conflicts()` in my latest commit.

Right now, I'm testing this since rebasing 12040 commits is not that fast.

And during the rebase I get some logs like this:
`dropping a4cec444eb3b158176f0650b414090197e7610d4 bump version to 0.8.2 -- patch contents already upstream`
and investigating if this is correct or not.

**Comment by adeebshihadeh** [comment]
Anything I can help with on this? Is there a particular blocker?

**Comment by andiradulescu** [comment]
The current blocker is `me`. Will try to unblock myself and make more progress today.

Otherwise, you can unlock the bounty. Maybe @devtekve is interested.

**Comment by devtekve** [comment]
> The current blocker is `me`. Will try to unblock myself and make more progress today.
> 
> Otherwise, you can unlock the bounty. Maybe @devtekve is interested.

Thanks ❤️ but not really, if it’s for me don’t unlock it! I was just chiming in. If there’s anything I can help with feel free to ping me tho

File: scripts/bfg.sh (hunk 1 above)
Comment (andiradulescu) [reply scripts/bfg.sh]: now you can find the latest analysis in this folder https://github.com/andiradulescu/openpilot/tree/bfg/scripts/analysis - I will try to import more files to lfs and I will keep updating the folder.

fyi `external` and `phonelibs` folder were initially deleted completely at the time of this comment https://github.com/commaai/openpilot/pull/31562#issuecomment-1961925056 - so still need to either add both folders to lfs or just add the big files inside them

File: scripts/bfg.sh (hunk 2 above)
Comment (andiradulescu) [reply scripts/bfg.sh]: I would also do it separately

File: scripts/bfg.sh (hunk 1 above)
Comment (andiradulescu) [reply scripts/bfg.sh]: analysis of the current commaai/openpilot repo can be found here https://github.com/andiradulescu/openpilot/tree/bfg/scripts/analysis-before

**Comment by andiradulescu** [comment]
With the latest `cherry-pick` script, the clone is `16.67 MB`:

image

But, I think I exaggerated a bit with LFS, please check [.gitattributes](https://github.com/andiradulescu/openpilot-tiny-cherry-pick/blob/master/.gitattributes). Most likely we can clean up the [git lfs import](https://github.com/andiradulescu/openpilot/blob/881c6c8bd6683c14bafccc453151010c10782524/scripts/rewrite-git-history-cherry-pick.sh#L127) and it will still be under `17 MB` or `18 MB`. What do you think we should remove (or add)?

**Comment by andiradulescu** [comment]
@adeebshihadeh I edited the first comment with some checkboxes - https://github.com/commaai/openpilot/pull/31562#issue-2151722885 - let me know what you think

**Comment by maxime-desroches** [comment]
Merging as is, feel free to open a pr if you need to add something to your script. We still have to wait untill all the cars stuff is out before actually doing the rewrite

**Comment by andiradulescu** [comment]
Glad seeing this merged! I’m still running it and changing small things to it, I’ll open a new PR after I stop testing and polishing.

**Pull request merged**